documentation/projects/proposals/trust_and_safety/content_report_moderation/20240122-implementation_plan_bulk_moderation_actions.md

```diff
@@ -0,0 +1,526 @@
+
+Using the join table to tie multiple records to a single decision (rather than,
+for example, creating a separate `ModerationDecision` record for each record
+selected for bulk update), has significant cost savings and more accurately
+models the data as a single decision.
+
+#### `action`
+
+This is a slug column describing the action taken. In addition to the action
+types defined in the related IP, we will add two new options:
+
+- `reversed_deindex` - Undo a previous deindexing decision.
+- `reversed_mark_sensitive` - Undo a previous decision marking the applicable
+  records sensitive.
+
+For consistency we will also update the existing action type
+`confirmed_sensitive` to `marked_sensitive` to describe the effect of the action
+without the use of the word "confirm", which is accurate when describing single
+decisions made by moderators but potentially inappropriate for describing bulk
+actions that do not include individual review of records.
+
+## Actions
+
+The related IP describes adding new custom Django
+[admin actions](https://docs.openverse.org/projects/proposals/trust_and_safety/content_report_moderation/20231208-implementation_plan_django_admin_moderator_access.html#actions)
+for each of the action types described in that IP. These actions are called from
+the Report view and take action on a list of MediaReports to create a
+ModerationDecision. They do not need to be modified, except to use the
+many-to-many `media_identifiers` column (although in this workflow only a single
+record will ever be added.)
+
+We will need to add new actions for making moderation decisions directly from a
+list of Media, as well as two new actions for the `reversed_<x>` action types.
+
+_All_ actions go through an intermediate confirmation page which:
+
+- Indicates the number of records that will be affected by the action
+- Collects the required `explanation` text (all other columns, such as
+  `moderator_id` and `created_on`, are filled in automatically)
+
+### Bulk Mark Sensitive: mark the works as sensitive
+
+This action takes a queryset of `AbstractMedia` records. It creates a
+ModerationDecision with type `marked_sensitive` and adds all records from the
+queryset **which do not already have a related `SensitiveMedia` record**. It
+then creates and saves `SensitiveMedia` records for each of these.
+
+The confirmation page for this action should indicate the number of records that
+will actually be affected. If some records are already marked sensitive, that
+should be indicated on the confirmation page as well.
+
+### Bulk Deindex sensitive: entirely exclude the works from Openverse due to sensitivity
+
+This action takes a queryset of `AbstractMedia` records. It creates a
+ModerationDecision with type `deindexed_sensitive` and adds all records from the
+queryset. It then creates and saves `DeletedMedia` records for each of these,
+deleting the record from the API and deindexing it in elasticsearch. The
```

Comment:
Currently when an individual record is deindexed, we do both (deleting the record from the API DB's main media table _and_ deindexing in Elasticsearch), so the intention was to apply consistent behavior to bulk actions.

This is an interesting idea, though. When a data refresh runs, records that have an associated DeletedMedia record are not copied over into the new media table or the index. So if we update the admin `deindex` logic to _only_ remove records from Elasticsearch, the record in the API media table will be retained _until the next data refresh_. It still gets removed from the table eventually, but we would have a built in "safety window" where it would be easier for us to reindex quickly, like you say 🤔 

We would:
* Update the `deindex` logic to only remove the record from Elasticsearch
* Make no changes to the data refresh logic, which will result in media with a `DeletedMedia` record being removed from the API media table on the next data refresh
* Update the `reindex` logic to check to see if the record being reindexed exists in the API media table, and if so we can immediately reindex it in elasticsearch

This would apply for bulk and individual actions. The cons I see are:
* The effect of a `reindex` action differs depending on when it is run, which may be confusing.
* A bulk `reindex` might be able to reindex _some_ but not _all_ of the selected records, if it affects records which were initially deindexed at different times :/
* The media tables and indices would become technically out of sync temporarily (and potentially for more than a week at a time if there are issues with the data refresh)

This is tough, bulk actions in particular reveal a lot of complexity.

Reply:
I had considered -- and @sarayourfriend also suggested :) -- putting a limit on the number of records that can be deindexed in any single decision, to mitigate risk in a much simpler way. I decided against it mainly because it is already possible to do large bulk `Delete` actions on media records, so I don't really perceive this as a _new_ risk.

Another thought is that once the data refresh is moved entirely into Airflow, I think it would be easy to add DAG conf options for a `SELECT` query that would allow you to run a "mini" data refresh updating only a subset of records more quickly (although I don't think we'll get around to that very soon).

Reply:
I have to say this is alot of complexity for what is currently a non-issue (quickly re-adding inappropriately deindexed data). It really only _becomes_ an issue if (a) someone takes a malicious action (which is too broad to try to prevent everything, at a certain level of access) or (b) someone mistakingly deindexes a disasterous number of documents.

Both of these are _exceptional_ circumstances. The easiest solution is to kick off an early data refresh, or accept that we will have _at most_ 7 days of bad data.

I really don't think we need contingencies beyond that, especially if they add so much complexity to an already complex set of issues. Adding a limit on the number of records deleted is much less complex and would be my preference for the _first_ thing we did _if_ someone accidentally deindexed a large chunk of works and for some reason a proactive data refresh proved not to be a suitable remedy.

Everything else feels excessively protectionist about data that is entirely recoverable. There's no risk of permanent loss, which makes me recoil from significant complexity to prevent temporarily deminished data caused by exceptional circumstances.

Reply:
Very strongly in agreement! I wanted to double check on this last option since as described it would be easy to _implement_ unlike some of the other ideas -- but it adds _conceptual_ complexity and edge cases that I personally did not like. It sounds like Olga is also okay with this so I'm going to go ahead and make revisions now.

Reply:
I'm okay making _no changes_ to prevent malicious or accidental deindexation of massive amounts of data until we have a reason to think it is not something from which we can easily recover within a week.

Reply:
I hadn't considered that some records could have been removed previously and not be in the API table. I agree with the reasoning that this is an exceptionally rare (and possibly malicious) case 👍